docs/DesignDocs/DeepCopyOutParameters.md

```diff
@@ -0,0 +1,838 @@
+
+    +   uint8_t* _deepcopy_out_buffer = NULL;
+    +   size_t _deepcopy_out_buffer_offset = 0;
+    +   size_t _deepcopy_out_buffer_size = 0;
+
+        size_t input_buffer_offset = 0;
+        size_t output_buffer_offset = 0;
+        OE_ADD_SIZE(input_buffer_offset, sizeof(*pargs_in));
+        OE_ADD_SIZE(output_buffer_offset, sizeof(*pargs_out));
+
+        /* Make sure input and output buffers lie within the enclave. */
+        /* oe_is_within_enclave explicitly checks if buffers are null or not. */
+        if (!oe_is_within_enclave(input_buffer, input_buffer_size))
+            goto done;
+
+        if (!oe_is_within_enclave(output_buffer, output_buffer_size))
+            goto done;
+
+        /* Set in and in-out pointers. */
+        /* There were no in nor in-out parameters. */
+
+        /* Set out and in-out pointers. */
+        /* In-out parameters are copied to output buffer. */
+        if (pargs_in->ptr)
+            OE_SET_OUT_POINTER(ptr, ((size_t)1 * sizeof(NestedBlob)), NestedBlob*);
+
+        /* Check that in/in-out strings are null terminated. */
+        /* There were no in nor in-out string parameters. */
+
+        /* lfence after checks. */
+        oe_lfence();
+
+        /* Call user function. */
+        foo(
+            pargs_in->ptr);
+
+    +   /* Compute the size for the deep-copy out buffer. */
+    +   if (pargs_in->ptr && pargs_in->ptr->blob_array)
+    +       OE_ADD_SIZE(_deepcopy_out_buffer_size, ((size_t)pargs_in->ptr->num * sizeof(Blob)));
+    +   for (size_t _i_2 = 0; _i_2 < pargs_in->ptr->num; _i_2++)
+    +   {
+    +       if (pargs_in->ptr && pargs_in->ptr->blob_array && pargs_in->ptr->blob_array[_i_2].buf)
+    +       OE_ADD_SIZE(_deepcopy_out_buffer_size, pargs_in->ptr->blob_array[_i_2].len);
+    +   }
+
+    +   if (_deepcopy_out_buffer_size)
+    +   {
+    +       _deepcopy_out_buffer = (uint8_t*) malloc(_deepcopy_out_buffer_size);
+    +       if (!_deepcopy_out_buffer)
+    +       {
+    +           _result = OE_OUT_OF_MEMORY;
+    +           goto done;
+    +       }
+    +   }
+
+    +   /* Serialize the deep-copied content into the buffer. */
+    +   if (pargs_in->ptr && pargs_in->ptr[_i_1].blob_array)
+    +       OE_WRITE_DEEPCOPY_OUT_PARAM(pargs_in->ptr[_i_1].blob_array, ((size_t)pargs_in->ptr[_i_1].num * sizeof(Blob)));
+    +   for (size_t _i_2 = 0; _i_2 < pargs_in->ptr[_i_1].num; _i_2++)
+    +   {
+    +       if (pargs_in->ptr && pargs_in->ptr[_i_1].blob_array && pargs_in->ptr[_i_1].blob_array[_i_2].buf)
+    +           OE_WRITE_DEEPCOPY_OUT_PARAM(pargs_in->ptr[_i_1].blob_array[_i_2].buf, pargs_in->ptr[_i_1].blob_array[_i_2].len);
+    +   }
+
+    +   if (_deepcopy_out_buffer_offset != _deepcopy_out_buffer_size)
+    +   {
+    +       _result = OE_FAILURE;
+    +       goto done;
+    +   }
+    +   OE_ADD_SIZE(output_buffer_offset, _deepcopy_out_buffer_size);
+
+    +   /* Allocate a host buffer and copy the content over to the buffer. */
```

Comment:
Instead of this, we could generate:
```
pargs_out->_host_buffer = _deepcopy_out_buffer;
pargs_out->_host_buffer_size = deepcopy_out_buffer_size;
```

The infrastructure can marshal this over to the host as appropriate. The generated code doesn't have to deal with host memory.

Reply:
Updated.